microraft-store-sqlite/src/main/java/io/microraft/store/sqlite/RaftSqliteStore.java

```diff
@@ -0,0 +1,346 @@
+
+import io.microraft.RaftEndpoint;
+import io.microraft.lifecycle.RaftNodeLifecycleAware;
+import io.microraft.model.RaftModelFactory;
+import io.microraft.model.log.LogEntry;
+import io.microraft.model.log.RaftGroupMembersView;
+import io.microraft.model.log.SnapshotChunk;
+import io.microraft.model.log.SnapshotEntry;
+import io.microraft.persistence.RaftStore;
+import io.microraft.persistence.RestoredRaftState;
+
+import java.io.File;
+import java.sql.Connection;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import org.jooq.CloseableDSLContext;
+import org.jooq.Converter;
+import org.jooq.Field;
+import org.jooq.Name;
+import org.jooq.Record;
+import org.jooq.Record1;
+import org.jooq.Select;
+import org.jooq.Table;
+import org.jooq.impl.DSL;
+import org.jooq.impl.SQLDataType;
+import org.sqlite.SQLiteConfig;
+import org.sqlite.SQLiteConfig.JournalMode;
+import org.sqlite.SQLiteConfig.LockingMode;
+import org.sqlite.SQLiteConfig.Pragma;
+
+/**
+ * An implementation of a RaftStore which uses SQLite for persistence. A user of this class is advised to construct
+ * this class, and then use {@link RaftSqliteStore#getRestoredRaftState()} to acquire any previously persisted state.
+ * <p>
+ * At time of writing, this store prioritizes:
+ * <ul>
+ *     <li>
+ *         Being reasonably low overhead on individual writes. It should not be a major source of
+ *         overhead when replicating messages.
+ *     </li>
+ *     <li>
+ *         Being as straightforward as possible with all other operations.
+ *     </li>
+ * </ul>
+ * There are three tables which exist in this store.
+ * <ol>
+ *     <li>logEntries stores the log entries</li>
+ *     <li>snapshots and snapshotChunks store which snapshots exist and pointers to their chunks</li>
+ *     <li>kv stores the remaining metadata in a single-row, one-column-per-value 'key-value store'.</li>
+ * </ol>
+ */
+public final class RaftSqliteStore implements RaftStore, RaftNodeLifecycleAware {
+    private static final Table<Record> KV = DSL.table("kv");
+
+    private static final Field<String> KEY = DSL.field("key", SQLDataType.VARCHAR);
+    private static final String PK = "pk";
+    private static final Field<Boolean> LOCAL_ENDPOINT_VOTING = DSL.field("localEndpointVoting", SQLDataType.BOOLEAN);
+    private static final Field<Integer> TERM = DSL.field("term", SQLDataType.INTEGER);
+    private static final Table<Record> LOG_ENTRIES = DSL.table("logEntries");
+    private static final Field<Long> INDEX = DSL.field("logIndex", SQLDataType.BIGINT);
+    private static final Table<Record> SNAPSHOT_CHUNKS = DSL.table("snapshotChunks");
+    private static final Field<Integer> CHUNK_INDEX = DSL.field("chunkIndex", SQLDataType.INTEGER);
+    private static final Table<Record> SNAPSHOTS = DSL.table("snapshots");
+    private static final Field<Integer> CHUNK_COUNT = DSL.field("chunkCount", SQLDataType.INTEGER);
```

Comment:
would it be easier if we use a single table to store snapshots? sth like `snapshotIndex, chunkCount, chunkIndex, chunkData`?

Reply:
done!